### PR TITLE
FOGL-9963 Add Alerts as a source for notifications

### PR DESCRIPTION
**Pull request by MarkRiddoch**
Provides a mechanism to deliver alerts via the notification system

### CODE REVIEW

**Review by pintomax — DISMISSED** [review]

File: C/services/notification/notification_subscription.cpp [hunk 1]
```diff
@@ -359,13 +407,17 @@ bool NotificationSubscription::addSubscription(SubscriptionElement *element)
...
 	m_subscriptions[key].push_back(element);
-	if (m_subscriptions[key].size() == 1)
+	if (m_subscriptions[key].size() <= 1)
 	{
 		if (element->registerSubscription(m_storage))
 			m_logger->info("Register for %s notification from the storage layer", key.c_str());
 		else
 			m_logger->error("Failed to register for %s notification from the storage layer", key.c_str());
 	}
+	else
```
Comment (praveen-garg) [inline C/services/notification/notification_subscription.cpp]: **Confusing else condition**:
The else is executed when m_subscriptions[key].size() > 1.

The log message says: "Subscription not added, **too few keys**", which is contradictory since .size() > 1 (else) implies more than one key.

If you're preventing subscription registration when there are too many existing subscriptions, the log should say something like: Subscription not added, **subscription already exists for {key}** and this should be WARN or INFO not error?

**Review by praveen-garg — COMMENTED** [review]
add the docs with how this help to end user and use cases?

and unit tests?

File: C/services/notification/notification_subscription.cpp (hunk 1 above)
Comment (MarkRiddoch) [reply C/services/notification/notification_subscription.cpp]: The entire else conditions was only meant to be there for debugging purposes and can be removed.

**Review by pintomax — DISMISSED** [review]

File: docs/index.rst [hunk 2]
```diff
@@ -103,6 +105,24 @@ posted with the audit log entry as the data points of the data. There is a
...
+
+The use of alerts as notification sources is however limited as these
+alerts are only capable of transporting a string to the notification
+system. This string describes the cause of the alert however. The primary
+use of alerts in notifications is to provide alternate channels for
+the alerts. Rather than simply showing the alert in the user interface
+menubar, the alert may be sent to any of the notification delivery
+channels. This greatly increases the ability to deliver these alerts
+to the consumers of alerts or end users not currently connected to the
+Fledge user interface.
```
Comment (praveen-garg) [inline docs/index.rst:133]: These two statements are not clear,

The primary use **of alerts** in notifications is to provide alternate channels for **the alerts**.

This greatly increases the ability to deliver these alerts to the consumers **of alerts** or end users not currently connected to the Fledge user interface.

File: docs/index.rst [hunk 3]
```diff
@@ -47,6 +47,8 @@ that adds an event engine to the Fledge installation. Notifications can be creat
 
   - The audit log entries that Fledge creates.
 
+  - The alerts raised by the Fledge instance.
```
Comment (praveen-garg) [inline docs/index.rst:56]: the connecting statement 

Notifications can be created based *on*

File: tests/unit/C/services/notification/DataAvailabilityRule_Test_Summary.md [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+- ✅ Plugin information retrieval
+- ✅ Builtin rule identification
+- ✅ Data persistence configuration
+
+#### 2. **Initialization and Configuration (100% covered)**
+- ✅ Empty configuration handling
+- ✅ Single audit code configuration
+- ✅ Single asset code configuration
+- ✅ Alerts configuration
+- ✅ Multiple comma-separated values
```
Comment (praveen-garg) [inline tests/unit/C/services/notification/DataAvailabilityRule_Test_Summary.md]: value of/for?

File: tests/unit/C/services/notification/DataAvailabilityRule_Test_Summary.md [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+}
+```
+
+## 📊 **Test Metrics**
+
+### Current Coverage Statistics
+- **Function Coverage**: 95%
+- **Branch Coverage**: 90%
+- **Line Coverage**: 92%
+- **Edge Case Coverage**: 85%
```
Comment (praveen-garg) [inline tests/unit/C/services/notification/DataAvailabilityRule_Test_Summary.md]: I would rather github ci workflow or Jenkins agent publish them to repo

File: tests/unit/C/services/notification/DataAvailabilityRule_Tests_Results.md [hunk 6]
```diff
@@ -0,0 +1,214 @@
+# DataAvailabilityRule Unit Tests - Results Summary
+
+## 🎉 **Test Execution Results**
+
+**Status: ✅ ALL TESTS PASSING**
+
+- **Total Tests**: 33
+- **Passed**: 33 ✅
+- **Failed**: 0 ❌
+- **Execution Time**: ~3ms per test run
```
Comment (praveen-garg) [inline tests/unit/C/services/notification/DataAvailabilityRule_Tests_Results.md]: I would rather github ci workflow or Jenkins agent publish them to repo.

File: C/services/notification/data_availability_rule.cpp [hunk 7]
```diff
@@ -147,6 +154,12 @@ string DataAvailabilityRule::triggers()
 		ret += "{ \"audit\" : \"" + audit + "\" }";
 		comma = ",";
 	}
+	if (m_alert)
+	{
+		ret += comma;
+		ret += "{ \"alert\" : \"alert\" }";
```
Comment (praveen-garg) [inline C/services/notification/data_availability_rule.cpp:160]: Why the hardcoded string value "alert", same as key?

In future, We better use `doc.AddMember` to avoid fragile and cumbersome string building!

File: C/services/notification/data_availability_rule.cpp [hunk 8]
```diff
@@ -45,6 +45,13 @@ static const char *default_config = QUOTE({
 		"default" : "",
 		"displayName" : "Asset Code",
 		"order" : "3"
+	},
+	"alerts" : {
+		"description" : "Notify when alerts are raised",
```
Comment (praveen-garg) [inline C/services/notification/data_availability_rule.cpp]: Probably, We should be clear here! 

Add alerts to the source of data for notifications
or
Notify for the alerts raised by the Fledge instance

File: tests/unit/C/services/notification/DataAvailabilityRule_Test_Summary.md (hunk 5 above)
Comment (MarkRiddoch) [reply tests/unit/C/services/notification/DataAvailabilityRule_Test_Summary.md]: Useful information from cursor, no harm it it and frankly I don't care ion it is here or not

**Comment by praveen-garg** [comment]
@MarkRiddoch should I expect a new commit? If no, please add comments to the conversation as appropriate so that I can close each of them and verify.

**Comment by MarkRiddoch** [comment]
> @MarkRiddoch should I expect a new commit? If no, please add comments to the conversation as appropriate so that I can close each of them and verify.

I accidentally hit the re-review on this PR. More commits are to follow.

File: C/services/notification/delivery_plugin.cpp [hunk 9]
```diff
@@ -188,3 +190,129 @@ void DeliveryPlugin::registerService(void *func, void *data)
...
+			message.c_str(), reason.c_str());
+	collectMacroInfo(rval, macros);
+	if (macros.size())
+	{
+		Document doc;
+		doc.Parse(reason.c_str());
+		if (doc.HasParseError())
+		{
+			// Failed to parse the reason, ignore macros
+			Logger::getLogger()->warn("Unable to parse reason document, macro substitutios withinthe notification will be ignored. The reason document is:  %s", reason.c_str());
```
Comment (praveen-garg) [inline C/services/notification/delivery_plugin.cpp]: space b/w within the

File: C/services/notification/delivery_plugin.cpp [hunk 10]
```diff
@@ -188,3 +190,129 @@ void DeliveryPlugin::registerService(void *func, void *data)
...
+		if (!doc.HasMember("data"))
+		{
+			Logger::getLogger()->warn("Unable to perform macro substitution in the notifcation alert. No data element was found in reason document %s", reason.c_str());
+			return rval;
+		}
+		Value& data = doc["data"];
+		Value::ConstMemberIterator itr = data.MemberBegin();
+		if (itr == data.MemberEnd())
+		{
+			Logger::getLogger()->warn("Unable to perform macro substitution in the notifcation alert. No data element has no children, reason document %s", reason.c_str());
```
Comment (praveen-garg) [inline C/services/notification/delivery_plugin.cpp]: No data element has no children?

File: C/services/notification/notification_subscription.cpp [hunk 11]
```diff
@@ -238,6 +238,54 @@ bool StatsRateSubscriptionElement::unregister(StorageClient& storage) const
...
+ *
+ * @param storage	The storage engine client
+ * @return bool		True if unregistered
+ */
+bool AlertSubscriptionElement::registerSubscription(StorageClient& storage) const
+{
+	NotificationApi *api = NotificationApi::getInstance();
+	string callBackURL = api->getAlertCallbackURL();
+	vector<std::string> keyValues;
+	Logger::getLogger()->fatal("Adding alert subscription for %s", callBackURL.c_str());
```
Comment (praveen-garg) [inline C/services/notification/notification_subscription.cpp]: why fatal?

File: C/services/notification/notification_api.cpp [hunk 12]
```diff
@@ -739,6 +797,48 @@ bool NotificationApi::queueStatsRateNotification(const string& statistic,
...
+	readingVec.push_back(reading);
+	ReadingSet* readings = NULL;
+	try
+	{
+		readings = new ReadingSet(&readingVec);
+	}
+	catch (exception* ex)
+	{
+		m_logger->error("Exception '" + string(ex->what()) + \
+				"' while parsing readings for alaert" + \
```
Comment (praveen-garg) [inline C/services/notification/notification_api.cpp]: alert typo

File: C/services/notification/data_availability_rule.cpp [hunk 13]
```diff
@@ -45,6 +45,13 @@ static const char *default_config = QUOTE({
 		"default" : "",
 		"displayName" : "Asset Code",
 		"order" : "3"
+	},
+	"alerts" : {
+		"description" : "Deliver alert data to the notificaiton delivery mechanism",
```
Comment (praveen-garg) [inline C/services/notification/data_availability_rule.cpp]: notificaiton typo

File: C/services/notification/delivery_plugin.cpp [hunk 14]
```diff
@@ -188,3 +190,129 @@ void DeliveryPlugin::registerService(void *func, void *data)
...
+ *
+ * @param message	The string to substitute into
+ * @param reason	The notification reason from which to pull data
+ * @return string	The substituted string
+ */
+string  DeliveryPlugin::expandMacros(const string& message, const string& reason)
+{
+	string rval = message;
+	vector<Macro> macros;
+	Logger::getLogger()->debug("Expand macros in messge %s with reason %s",
```
Comment (praveen-garg) [inline C/services/notification/delivery_plugin.cpp]: messge typo

**Review by praveen-garg — COMMENTED** [review]
You may want to check #107

File: C/services/notification/notification_api.cpp [hunk 15]
```diff
@@ -739,6 +797,48 @@ bool NotificationApi::queueStatsRateNotification(const string& statistic,
...
+
+	Reading *reading = new Reading("alert", payload);
+	vector<Reading *> readingVec;
+	readingVec.push_back(reading);
+	ReadingSet* readings = NULL;
+	try
+	{
+		readings = new ReadingSet(&readingVec);
+	}
+	catch (exception* ex)
```
Comment (gnandan) [inline C/services/notification/notification_api.cpp:818]: Constructor of ReadingSet thows exception of type ReadingSetException. Please catch by ReadingSetException* to align with  ReadingSet class

File: C/services/notification/notification_api.cpp [hunk 16]
```diff
@@ -567,6 +584,47 @@ void NotificationApi::processStatsRateCallback(shared_ptr<HttpServer::Response>
...
+		}
+		else
+		{
+			responsePayload = "{ \"error\": \"error_message\" }";
+			this->respond(response,
+				      SimpleWeb::StatusCode::client_error_bad_request,
+				      responsePayload);
+		}
+	}
+	catch (exception ex)
```
Comment (gnandan) [inline C/services/notification/notification_api.cpp:622]: Catching  exception by value? consider catching by reference with is safer option.

File: C/services/notification/notification_subscription.cpp [hunk 17]
```diff
@@ -238,6 +238,54 @@ bool StatsRateSubscriptionElement::unregister(StorageClient& storage) const
...
+ */
+bool AlertSubscriptionElement::registerSubscription(StorageClient& storage) const
+{
+	NotificationApi *api = NotificationApi::getInstance();
+	string callBackURL = api->getAlertCallbackURL();
+	vector<std::string> keyValues;
+	Logger::getLogger()->info("Adding alert subscription for %s", callBackURL.c_str());
+	if (!storage.registerTableNotification("alerts", "", keyValues, "insert", callBackURL))
+		Logger::getLogger()->error("Failed to register insert handler for alert subscription");
+	return storage.registerTableNotification("alerts", "", keyValues, "update", callBackURL);
```
Comment (gnandan) [inline C/services/notification/notification_subscription.cpp:271]: consider adding a debug message to log if registerSubscription was successful.

File: C/services/notification/notification_subscription.cpp [hunk 18]
```diff
@@ -238,6 +238,54 @@ bool StatsRateSubscriptionElement::unregister(StorageClient& storage) const
...
+ * @param storage	The storage engine client
+ * @return bool		True if unregistered
+ */
+bool AlertSubscriptionElement::unregister(StorageClient& storage) const
+{
+	NotificationApi *api = NotificationApi::getInstance();
+	string callBackURL = api->getStatsRateCallbackURL();
+	vector<std::string> keyValues;
+	storage.unregisterTableNotification("alerts", "", keyValues, "update", callBackURL);
+	return storage.unregisterTableNotification("alerts", "", keyValues, "insert", callBackURL);
```
Comment (gnandan) [inline C/services/notification/notification_subscription.cpp:286]: consider adding a debug message to log if unregister was successful.

File: docs/index.rst [hunk 19]
```diff
@@ -103,6 +111,27 @@ posted with the audit log entry as the data points of the data. There is a
 limited set of notification rule plugins that can be used with this data as
 it tends to be non-numeric and most plugins expect to sue numeric data.
 
+Alerts
+------
+
+Fledge will alert users to specific actions using the *bell* icon on
+the menu bar. These alerts can be used as a source of notification data
+by some of the notification plugins. Most notably the data availability
```
Comment (gnandan) [inline docs/index.rst:119]: As JIRA mentions about match rules, can we include **foglamp-rule-match** plugin as well along with data availability here or somewhere else?

File: tests/unit/C/services/notification/test_async_config_change.cpp [hunk 20]
```diff
@@ -18,6 +18,7 @@ class AsyncConfigChangeTest : public ::testing::Test
 protected:
     void SetUp() override
     {
+        delete Logger::getLogger();
```
Comment (gnandan) [inline tests/unit/C/services/notification/test_async_config_change.cpp:21]: Why Logger is deleted? Where it is used ?

File: tests/unit/C/services/notification/test_data_availability_rule.cpp [hunk 21]
```diff
@@ -0,0 +1,771 @@
...
+ * Test evaluation with multiple audit codes
+ */
+TEST_F(DataAvailabilityRuleTest, EvalWithMultipleAuditCodes)
+{
+    // Arrange
+    DataAvailabilityRule rule("TestDataAvailability");
+    ConfigCategory config = createBasicConfig("AUDIT001,AUDIT002");
+    rule.init(config);
+    
+    string testJSON = "{\"AUDIT001\": \"value1\", \"AUDIT002\": \"value2\"}";
```
Comment (gnandan) [inline tests/unit/C/services/notification/test_data_availability_rule.cpp:423]: Can raw string literal with R"()" be used here?

File: tests/unit/C/services/notification/test_delivery_plugin_expand_macros.cpp [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    DeliveryPlugin* m_plugin;
+};
+
+// Test basic macro substitution with string values
+TEST_F(DeliveryPluginExpandMacrosTest, BasicStringMacroSubstitution)
+{
+    string message = "Temperature is $temperature$ degrees";
+    string reason = "{\"reason\": \"triggered\", \"auditCode\": \"test\", \"data\": { \"example\" : {\"temperature\": \"25.5\"}}}";
```
Comment (gnandan) [inline tests/unit/C/services/notification/test_delivery_plugin_expand_macros.cpp:31]: can raw string literal with R"()" be used ?

File: C/services/notification/notification_manager.cpp [hunk 23]
```diff
@@ -1230,14 +1230,17 @@ bool NotificationManager::APIcreateEmptyInstance(const string& name)
 			 "\"enumeration\", \"options\": [ \"one shot\", \"retriggered\", \"toggled\" ], "
 			 "\"displayName\" : \"Type\", \"order\" : \"4\","
 			 "\"default\" : \"one shot\"}, "
+		   "\"text\": {\"description\": \"Text message to send for this notification\", "
```
Comment (gnandan) [inline C/services/notification/notification_manager.cpp:1233]: Option to provide value for message is not available during creating Notification instance. It appears later on the GUI once Notification instance is created. It leads to confusion. Please rectify it.

Comment (MarkRiddoch) [reply C/services/notification/notification_manager.cpp:1233]: Yes, this is a known issue in the GUI and will be address in the GUI

File: tests/unit/C/services/notification/test_delivery_plugin_expand_macros.cpp (hunk 22 above)
Comment (MarkRiddoch) [reply tests/unit/C/services/notification/test_delivery_plugin_expand_macros.cpp:31]: It could on those platforms that support it. We don't use it generally because we had issues on some platforms. Hence the use of QUOTE instead.

File: tests/unit/C/services/notification/test_data_availability_rule.cpp (hunk 21 above)
Comment (MarkRiddoch) [reply tests/unit/C/services/notification/test_data_availability_rule.cpp:423]: It could on those platforms that support it. We don't use it generally because we had issues on some platforms. Hence the use of QUOTE instead.

File: tests/unit/C/services/notification/test_async_config_change.cpp (hunk 20 above)
Comment (MarkRiddoch) [reply tests/unit/C/services/notification/test_async_config_change.cpp:21]: It gets used in some of the routines that are called by the tests and hence needs to be deleted.

File: docs/index.rst (hunk 19 above)
Comment (MarkRiddoch) [reply docs/index.rst:119]: That is in a different repository and is therefore not part of this pull request.

File: C/services/notification/notification_subscription.cpp (hunk 18 above)
Comment (MarkRiddoch) [reply C/services/notification/notification_subscription.cpp:286]: The Jira is to do a specific task, not to make general changes, Therefore making unrelated changes is not part of this work. If a bug was spotted then it should be fixed, tis does not fall into that category.

File: C/services/notification/notification_subscription.cpp (hunk 17 above)
Comment (MarkRiddoch) [reply C/services/notification/notification_subscription.cpp:271]: The Jira is to do a specific task, not to make general changes, Therefore making unrelated changes is not part of this work. If a bug was spotted then it should be fixed, tis does not fall into that category.

File: C/services/notification/notification_api.cpp (hunk 16 above)
Comment (MarkRiddoch) [reply C/services/notification/notification_api.cpp:622]: The Jira is to do a specific task, not to make general changes, Therefore making unrelated changes is not part of this work. If a bug was spotted then it should be fixed, tis does not fall into that category.

File: C/services/notification/notification_api.cpp (hunk 15 above)
Comment (MarkRiddoch) [reply C/services/notification/notification_api.cpp:818]: The Jira is to do a specific task, not to make general changes, Therefore making unrelated changes is not part of this work. If a bug was spotted then it should be fixed, tis does not fall into that category.

**Review by pintomax — APPROVED** [review]

File: C/services/notification/notification_manager.cpp (hunk 23 above)
Comment (praveen-garg) [reply C/services/notification/notification_manager.cpp:1233]: Can we please log this issue?

**Review by gnandan — APPROVED** [review]
Reviewed by testing. Data Availability rule allows to add alerts as a source.

File: tests/unit/C/services/notification/test_data_availability_rule.cpp (hunk 21 above)
Comment (gnandan) [reply tests/unit/C/services/notification/test_data_availability_rule.cpp:423]: No QUOTE has been used.

File: tests/unit/C/services/notification/test_delivery_plugin_expand_macros.cpp (hunk 22 above)
Comment (gnandan) [reply tests/unit/C/services/notification/test_delivery_plugin_expand_macros.cpp:31]: No QUOTE has been used.

File: C/services/notification/notification_manager.cpp (hunk 23 above)
Comment (gnandan) [reply C/services/notification/notification_manager.cpp:1233]: @praveen-garg following JIRA has been logged https://scaledb.atlassian.net/browse/FOGL-10239
